src/test/java/com/fasterxml/jackson/databind/cfg/CacheProviderTest.java

```diff
@@ -0,0 +1,182 @@
+    
+    @Before
+    public void setUp() {
+        createdCaches.clear();
+        invoked = false;
+    }
+
+    @Test
+    public void testDefaultCacheProviderConfigDeserializerCache() throws Exception
+    {
+        CacheProvider cacheProvider = DefaultCacheProvider.builder()
+                .maxDeserializerCacheSize(1234)
+                .build();
+        ObjectMapper mapper = JsonMapper.builder()
+                .cacheProvider(cacheProvider).build();
+
+        assertNotNull(mapper.readValue("{\"point\":24}", RandomBean.class));
+    }
+
+    @Test
+    public void testDefaultCacheProviderConfigDeserializerCacheSizeZero() throws Exception
+    {
+        CacheProvider cacheProvider = DefaultCacheProvider.builder()
+                .maxDeserializerCacheSize(0)
+                .build();
+        ObjectMapper mapper = JsonMapper.builder()
+                .cacheProvider(cacheProvider)
+                .build();
+
+        assertNotNull(mapper.readValue("{\"point\":24}", RandomBean.class));
+    }
+
+    @Test
+    public void testBuilderNullCheckingForDeserializerCacheConfig() throws Exception
+    {
+        try {
+            DefaultCacheProvider.builder()
+                    .maxDeserializerCacheSize(-1);
+            fail("Should not reach here");
+        } catch (IllegalArgumentException e) {
+            assertTrue(e.getMessage().contains("Cannot set maxDeserializerCacheSize to a negative value"));
+        }
+    }
+
+    @Test
+    public void testCustomCacheProviderConfig() throws Exception
+    {
+        ObjectMapper mapper = JsonMapper.builder()
+                .cacheProvider(new CustomCacheProvider())
+                .build();
+
+        assertTrue(invoked);
+        assertNotNull(mapper.readValue("{\"point\":24}", RandomBean.class));
+    }
+
+    @Test
+    public void testDefaultCacheProviderSharesCache() throws Exception
+    {
+        // Arrange
+        // 1. shared CacheProvider
+        CacheProvider cacheProvider = new CustomCacheProvider();
+        
+        // 2. two different mapper instances
+        ObjectMapper mapper1 = JsonMapper.builder()
+                .cacheProvider(cacheProvider)
+                .build();
+        ObjectMapper mapper2 = JsonMapper.builder()
+                .cacheProvider(cacheProvider)
+                .build();
+        
+        // Act 
+        // 3. Add two different types to each mapper cache
+        mapper1.readValue("{\"point\":24}", RandomBean.class);
+        mapper2.readValue("{\"height\":24}", AnotherBean.class);
+        
+        // Assert
+        // 4. Should have created two cache instance
+        assertEquals(2, createdCaches.size());
+    }
+
+    @Test
+    public void testBuilderBuildWithDefaults() throws Exception
+    {
+        try {
+            DefaultCacheProvider.builder().build();
+        } catch (IllegalArgumentException e) {
+            fail("Should not reach here");
```

Comment:
I don't think this is needed -- let the exception be thrown if it is; that'll fail test just fine.